jobs/MeterInstanceJob.js

```diff
@@ -0,0 +1,185 @@
+        .catch((err) => this.runFailed(err, {}, job, done));
+    });
+  }
+
+  static getInstanceEvents() {
+    const options = {
+      resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.INSTANCE,
+      resourceType: CONST.APISERVER.RESOURCE_TYPES.SFEVENT,
+      query: {
+        labelSelector: `meter_state=${CONST.METER_STATE.TO_BE_METERED}`
+      }
+    };
+    return apiServerClient.getResources(options);
+  }
+
+  static meter(events) {
+    logger.info(`Number of events to be metered in this run - ${events.length}`);
+    return Promise.try(() => {
+      let successCount = 0,
+        failureCount = 0,
+        failedEvents = [];
+      return Promise.map(events, (event) => {
+          return this
+            .sendEvent(event)
+            .then((status) => {
+              if (status) {
+                successCount++;
+              } else {
+                failureCount++;
+                failedEvents.push(event);
+              }
+            })
+            .catch(err => {
+              logger.error(`Error occurred while metering event : `, err);
+            });
+        }).then(() => {
+          return {
+            totalEvents: events.length,
+            success: successCount,
+            failed: failureCount,
+            failedEvents: failedEvents
+          };
+        })
+        .catch(err => {
+          logger.error(`Error occurred while metering all events : `, err);
+        });
+
+    });
+  }
+
+  static isServicePlanExcluded(options) {
+    options = JSON.parse(options);
+    const serviceId = _.get(options, 'service.id');
+    const planId = _.get(options, 'service.plan');
+    logger.info(`Checking if service ${serviceId}, plan: ${planId} is excluded`);
+    const serviceName = this.getServiceNameFromServiceGUID(serviceId);
+    const planName = this.getPlanSKUFromPlanGUID(serviceId, planId);
+    const excluded_service_names = _.map(config.metering.excluded_service_plans, p => p.service_name);
+    if (_.indexOf(excluded_service_names, serviceName) >= 0) {
+      const excluded_plans = _
+        .chain(config.metering.excluded_service_plans)
+        .filter(p => p.service_name === serviceName)
+        .head()
+        .get('plan_sku_names')
+        .value();
+      logger.info(`Excluded plans for ${serviceName}:`, excluded_plans);
+      if (_.indexOf(excluded_plans, planName) >= 0) {
+        logger.info(`The metering event guid: ${options.id} with ${serviceName}, ${planName} is excluded`);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static enrichEvent(options) {
+    // Add region , service name and plan sku name of the event
+    options = JSON.parse(options);
+    const serviceId = _.get(options, 'service.id');
+    const planId = _.get(options, 'service.plan');
+    logger.info(`Enriching the metering event ${serviceId}, plan: ${planId}`);
+    options.service.id = this.getServiceNameFromServiceGUID(serviceId);
```

Comment:
Also Can't you use directly https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/common/models/catalog.js#L37 for the desired info.

Reply:
Please see: 
https://wiki.wdf.sap.corp/wiki/display/CPC15N/Usage+Document+Detailed+Schema
The usage send to MaaS has object:
```
service (object)
 - id  // the name with which the service is registered eg postgresql, redis etc
 - plan // sku name 
```
We populate the guid in webhook and enrich it in the broker before sending.

Reply:
ok , they have used slightly confusing names..

Just a question/suggestion : With this then we are tying SF to SAP based metering ? Can we introduce some flexibility here such that we can have some templates and we populate the data based on the same ? Does it make sense ? Or will Metering2.0 be opensource standard for metering service instances..

Reply:
sure, now since we have Interoperator,  we could have metering data as part of the sfInstance CRD (part of interoperator) and then  interoperator can make metering specific CRDs . This way any other metering service can bring their templates for interoperator and use the metering data already present in SFInstance.

@subhankarc Does the above approach make sense ?